### DISCUSSION TITLE
How to use Prettier with tsserver?

### DISCUSSION THREAD

**Discussion by teenjuna** [mason-org/mason.nvim#320] [Q&A]
Hi! I'm trying to use `Prettier` along with `typescript-language-server`. I have installed both, but there's no sight of prettier anywhere. Right now I'm not `require` `null-ls` anywhere. 

This is how I setup `Mason`:
```lua
use {
	'williamboman/mason.nvim',
	requires = {
		{ 'williamboman/mason-lspconfig.nvim' },
		{ 'neovim/nvim-lspconfig' },
		{ 'simrat39/rust-tools.nvim' },
	},
	config = function()
		require('mason').setup()
		require('mason-lspconfig').setup()
		require('rust-tools').setup()
	end,
}
```

And this is how I connect it with `nvim-cmp`:
```lua
-- ...
local capabilities = require('cmp_nvim_lsp').update_capabilities(vim.lsp.protocol.make_client_capabilities())
require('mason-lspconfig').setup_handlers {
	function(server_name)
		require('lspconfig')[server_name].setup {
			capabilities = capabilities,
			handlers = handlers
		}
	end,
	['cssls'] = function()
		require('lspconfig').cssls.setup {
			capabilities = capabilities,
			handlers = handlers,
			settings = {
				css = {
					lint = {
						unknownAtRules = 'ignore',
					}
				}
			}
		}
	end
}
```

What should I add?

**Comment by williamboman** [comment]
Hello! Hm what do you mean by no sight of prettier anywhere? To my knowledge, typescript-language-server does not have any kind of integration with prettier.

Try `:!prettier --write %`, or use a linter/formatter plugin such as null-ls for a more in-depth integration.

**Reply by teenjuna** [reply]
Let me clarify. Previously I used the lspconfig manually, and I remember that in order to make it work like I want (typescript-language-server for lsp info, prettier for formatting) I needed to setup typescript-language-server with disabled formatting capability, and setup null-ls with prettier integration). I think I need to do the same thing here, but then I'm not sure what installing Prettier through Mason really does (like, does it setup null-ls at all?). I guess I'm confused

**Reply by williamboman** [reply]
Mason doesn't set up any other plugins for you. It installs external tooling (such as prettier) and make them available within Neovim. Generally speaking, what gets installed by Mason is one or more executables, per package. How you use those executables is entirely up to you. `!prettier --write %` is one method. Using `null-ls` is another method. Using `formatter.nvim` is a third. Writing your own solution that suits your needs is a fourth. The fifth method would be to do exactly what you did before, Mason doesn't replace any of the workflows or plugins you mentioned, it solves different problems. The possibilities are endless.
